server/src/main/java/org/elasticsearch/common/io/stream/DelayableWriteable.java

```diff
@@ -136,20 +156,20 @@ public void writeTo(StreamOutput out) throws IOException {
                  * differences in the wire protocol. This ain't efficient but
                  * it should be quite rare.
                  */
-                referencing(get()).writeTo(out);
+                referencing(expand()).writeTo(out);
             }
         }
 
         @Override
-        public T get() {
+        public T expand() {
             try {
                 try (StreamInput in = registry == null ?
                         serialized.streamInput() : new NamedWriteableAwareStreamInput(serialized.streamInput(), registry)) {
```

Comment:
I was wondering, is it an option to keep a reference to the original stream input so that we do not need this if and we can avoid exposing the new namedWriteableRegistry getter in StreamInput?

Reply:
I think we could keep a reference to the original `StreamInput` but we'd still need to read the serialized bytes from it.

I wonder if keeping a reference to the `StreamInput` would keep a reference to its underlying storage. If it did then it'd probably be bad to keep the reference.

Reply:
yea it did sound risky to me as well, and what we do instead looks fine to me.